sklearn/ensemble/_hist_gradient_boosting/_loss.pyx

```diff
@@ -153,11 +156,12 @@ def _update_gradients_hessians_categorical_crossentropy(
         int n_samples = raw_predictions.shape[1]
         int k  # class index
         int i  # sample index
-        Y_DTYPE_C sw
+        G_H_DTYPE_C sw
```

Comment:
why change sw?

Reply:
Otherwise, in line 187-188, `gradients[k, i] = XXX * sw`, the multiplication on the right-hand side would be a double, which is then down-casted to float32 for assignment to the left-hand side.
Why doing the multiplication in double precision, when it's down-casted to float anyway?

Reply:
>  Why doing the multiplication in double precision, when it's down-casted to float anyway?


To avoid too many casts since casting isn't free. Though I don't know enough of the details to understand how costly they are, and how/when exactly they happen.

Note that according to this logic, y_true in `_update_gradients_hessians_categorical_crossentropy` should be converted to float too since we do  `gradients[i] = p_i - y_true[i]` (I don't think we should, just noting a discrepancy.)

Reply:
`y_true` is a memory view, i.e. `double[:]`. I guess there is no implicit type conversion (possible). Accoding to [this answer](https://stackoverflow.com/questions/37432076/cython-typed-memoryviews-what-they-really-are/37497998#37497998), the underlying type is a struct containing pointers. Providing a float32 array as argument throws an error instead of casting.